server/apps/migration/core-data-jpa-to-pg/src/main/java/org/apache/james/JpaToPgCoreDataMigration.java

```diff
@@ -0,0 +1,455 @@
+                                    var jpaUser = (JPAUser) user;
+                                    var pgUser = new DefaultUser(
+                                            username,
+                                            jpaUser.getPasswordHash(),
+                                            jpaUser.getAlgorithm(),
+                                            algorithm
+                                    );
+                                    try {
+                                        postgresUsersDAO.addUser(username, UUID.randomUUID().toString());
+                                    } catch (RuntimeException e) {
+                                        if (!(e.getCause() instanceof AlreadyExistInUsersRepositoryException)) {
+                                            throw e;
+                                        }
+                                        // Idempotent behavior, if the user already exists, we
+                                        // only update it.
+                                    }
+                                    Throwing.runnable(() ->
+                                            postgresUsersDAO.updateUser(pgUser)
+                                    ).sneakyThrow().run();
+
+                                });
+                            } catch (UsersRepositoryException e) {
+                                LOGGER.warn("Failed to migrate user {}", username, e);
+                            }
+                        }
+                );
+            } catch (UsersRepositoryException e) {
+                throw new RuntimeException("Failed to retrieve users for migration", e);
+            }
+            LOGGER.info("Users migration complete");
+        }
+    }
+
+    static class RRTMigration {
+        private static final Logger LOGGER = LoggerFactory.getLogger(RRTMigration.class);
+
+        private final JPARecipientRewriteTable jpaRRTRepository;
+        private final PostgresRecipientRewriteTable pgRRTRepository;
+
+        @Inject
+        RRTMigration(
+                JPARecipientRewriteTable jpaRRTRepository,
+                PostgresRecipientRewriteTable pgRRTRepository
+        ) {
+            this.jpaRRTRepository = jpaRRTRepository;
+            this.pgRRTRepository = pgRRTRepository;
+        }
+
+        void doMigration() {
+            LOGGER.info("Start RRT migration");
+            try {
+                jpaRRTRepository.getAllMappings().forEach((mappingSource, mappings) ->
+                        mappings.forEach(mapping ->
+                                pgRRTRepository.addMapping(mappingSource, mapping)
+                        )
+                );
+            } catch (RecipientRewriteTableException e) {
+                throw new RuntimeException(e);
+            }
+            LOGGER.info("RRT migration complete");
+        }
+    }
+
+    static class MailRepositoryUrlMigration {
```

Comment:
IMO no need to migrate this as it would be created by the app upon startup.

Reply:
>What really is valuable however is migrating Mail Repositories which do not seems achieved here.
indeed not, I was blinded by my bias of building a tool to migrate scaling-pulsar-smtp which doesn't use the SQL database to store mails only the URLs :D I'll look into adding that, but i must say one would have to be careful with the volume as the migration will effectively copy all the data from one table to the other.

Reply:
I pushed a fixup that adds MailRepository migration. The fixup will not work with deduplication. The deduplication strategy [forcibly installs a webmin route](https://github.com/apache/james-project/blob/master/server/container/guice/blob/deduplication-gc/src/main/java/org/apache/james/modules/blobstore/BlobDeduplicationGCModule.java#L67) module which makes absolutely no sense for the migration tool and I want to make progress on Multi-DKIM storage and ARC instead of fighting that feature. 

It does work with passthrough as demonstrated by the test.